fix queue stats to included 'processed' tasks

elv-gilles opened the pull request:
* add state for processed tasks
* add function in inspector to delete all processed tasks
* add synchronization on queues
* fix go mod in tools sub package

In the redis implementation processed keys are removed from the tasks set and a counter is incremented under a key of the form `asynq:{queue_name}processed:2006-01-02`. The same for failed tasks.

The current incarnation for [r/s]qlite instead keeps all tasks in the tasks table.
We may change it and implement something closer to the redis one, but - for now, as a matter of simplicity - I just fixed the queue stats and added a function `DeleteAllProcessedTasks` to the inspector to allow cleaning up.

elv-gilles committed:
fix queue stats to included 'processed' tasks

* add state for processed tasks
* add function in inspector to delete all processed tasks
* add synchronization on queues
* fix go mod in tools sub package

lukseven commented on queues.go:
```diff
@@ -143,6 +154,14 @@ func (p *QueuesConfig) Names() []string {
 	return uniq(names, len(p.queueConfig))
 }
 
+func (p *QueuesConfig) UpdateQueues(queues map[string]int) error {
```
I don't see this function called anywhere.

lukseven reviewed (approved):
As discussed previously, I think keeping stats is not a core function of Asynq. Instead, we have to make sure that we get rid of garbage. So we should either remove processed tasks as soon as they enter that state, or have a background cleanup mechanism that calls DeleteAllProcessedTasks periodically. I would personally opt for the former. Can be implemented in a follow-up PR.

jslching commented:
Thanks, what is the functional difference between `TaskStateCompleted` and `TaskStateProcessed`?

elv-gilles committed:
small fixes & state diagram update

* processor: missing release token on failure
* asynq: missing task state for processed tasks
* listTasksPaged: order by insertion order by default

elv-gilles committed:
add constructor of Client for sharing the broker of a server

elv-gilles committed:
add list option in inspector and fixes

* inspector: add option start_after_uuid (r/sqlite only)
* client: round timeout of tasks to second (bug)
* gorqlite: use updated version
* fix array index outof bound in schema & tasks
* fix unmerged change in cmd

elv-gilles committed:
fix requeuing, add initial check for healthchecker, recoverer and al.

* fix (r/sqlite):
  * requeuing was preventing tasks to execute (server-id was sticky present)
  * retryTask was not setting the deadline to NULL
* healthchecker:
  * do initial check when starting instead of waiting the interval
  * implement initial check for recoverer, forwarder & janitor relying on
    notification of good health dispatched by healthchecker instead of poll interval
* janitor: made interval configurable
* misc: fixed wrong log formatting

elv-gilles committed:
fix array index out of bounds, rqlite returning no error and no result.

* when cluster shutting down

elv-gilles committed:
update gorqlite, add func listOption.pagination

elv-gilles commented:
> Thanks, what is the functional difference between `TaskStateCompleted` and `TaskStateProcessed`?

None. Initially there were only `TaskStateProcessed` (and they were removed when done). Now there's also `TaskStateCompleted` by which tasks are subject to retention.

elv-gilles commented:
> remove processed tasks as soon as they enter that state

I'll make another PR to remove them.

elv-gilles committed:
add end-2-end test TestCancelProcessing

* avoid race on processor.wait
* minor typos and logs

elv-gilles committed:
avoid race on processor.wait

* WARNING: DATA RACE
  Write at 0x00c000073a50 by goroutine 17:
    runtime.closechan()
        /opt/hostedtoolcache/go/1.17.13/x64/src/runtime/chan.go:355 +0x0
    github.com/hibiken/asynq.(*processor).shutdownNow()
        /home/runner/work/asynq/asynq/processor.go:168 +0x204
    github.com/hibiken/asynq.(*Server).shutdown()
  Previous read at 0x00c000073a50 by goroutine 78:
    runtime.chansend()
        /opt/hostedtoolcache/go/1.17.13/x64/src/runtime/chan.go:158 +0x0
    github.com/hibiken/asynq.(*processor).fini.func4.1()
        /home/runner/work/asynq/asynq/processor.go:301 +0x437

elv-gilles committed:
require go 1.18, update github workflows

elv-gilles committed:
go mod tidy

elv-gilles committed:
require go 1.19

elv-gilles committed:
avoid race on processor.wait (2)

elv-gilles committed:
fix another race in tests

elv-gilles committed:
fix (?) test on CI build

elv-gilles committed:
skip test

elv-gilles committed:
upgrade eluv-io/gorqlite to v0.0.7
